content/articles/automating-jobs-scheduling-with-django_cron-in-python-application/index.md

```diff
@@ -0,0 +1,395 @@
+
+```bash
+python -m venv env
+source env/Scripts/activate
+
+pip install django django_cron djangorestframework
+pip freeze > requirements.txt
+```
+
+You have configured the environment for the project, and installed all of the dependencies. You are keeping track of them via the `requirements.txt` file.
+
+Let's start the project, spin up the Django server and create an app called `news` with the commands below:
+
+```bash
+django-admin startproject hackernewsdemo
+python manage.py startapp news
+
+python manage.py runserver
+```
+
+If you open your browser at <http://127.0.0.1:8000> you should see the default Django welcome page as shown below:
+
+![First page](/engineering-education/automating-jobs-scheduling-with-django_cron-in-python-application/first-page.png)
+
+Now open to the project `settings.py` file, navigate to the `INSTALLED_APPS`, and add the snippet below:
+
+```python
+# ...
+'rest_framework',
+'django_cron',
+'news'
+```
+
+In addition, it is mandatory to register the `CRON_CLASSES` list in the `settings.py` right below the file. The list contains all the cron classes that the application has. Also, we need to add the permission classes for the API. 
+
+```python
+CRON_CLASSES = [
+    "news.cron.MyCronJob",
+    # ...
+]
+
+REST_FRAMEWORK = {
+  'DEFAULT_PERMISSION_CLASSES': [
+      'rest_framework.permissions.AllowAny',
+  ],
+}
+```
+
+#### Making models for the endpoint
+In this API, what is actually needed from the external endpoint is the `ids` of the external news data. For this reason, we will make the model with just two properties which are the `id` and the `datetime` fetched.
+
+It is important to make the primary key of the database from the fetched data `id`. This is because it will be referenced whenever you are to get the latest news from the external endpoint.
+
+Add the snippet below into your `models.py` of the `news` app.
+
+```python
+from django.db import models
+from datetime import datetime
+
+class HackerNewsID(models.Model):
+    hackernews = models.BigIntegerField(unique=True, primary_key=True)
+    fetched_at = models.DateTimeField(default=datetime.now())
+
+    def save(self, *args, **kwargs):
+        self.id = self.hackernews # replacing the id(primary key) as the hackernews id
+        super(HackerNewsID, self).save(*args, **kwargs)
+
+    def __str__(self):
+        return str(self.hackernews)
+```
+
+- Note: The `save` method under the class `HackerNewsID` will be called whenever data is been saved. This only formats the primary key to the `id` fetched.
+
+#### Making cron jobs
+Here, we will make the cron job class. This class is required by the `django_cron` library to run every minute(s) as configured in the code below.
+
+The code snippet below describes the cron job class together with the method that will be called whenever the command below is executed.
+
+```bash
+python manage.py runcrons
+```
+--- 
+
+```python
+from django_cron import CronJobBase, Schedule
+import requests
```

Comment:
@damikanye  Please make it clear to which file this code is going into. Make sure all your code snippets have the same instructions.

Reply:
Resolved.